internal/manifest/oci_index.go

```diff
@@ -73,15 +83,78 @@ func (index *OCI1IndexPublic) UpdateInstances(updates []ListUpdate) error {
 	return nil
 }
 
-// ChooseInstance parses blob as an oci v1 manifest index, and returns the digest
-// of the image which is appropriate for the current environment.
-func (index *OCI1IndexPublic) ChooseInstance(ctx *types.SystemContext) (digest.Digest, error) {
+// Compares manifest and bestMatched with each other and returns true if manifest has preference over bestMatched,
+// otherwise returns false.
+func instanceBetterMatchesCompression(manifest imgspecv1.Descriptor, bestMatched imgspecv1.Descriptor, didPreferGzip bool) bool {
+	if bestMatched.Digest == "" {
+		return true
+	}
+
+	bestMatchIsZstd := instanceIsZstd(bestMatched)
+	manifestIsZstd := instanceIsZstd(manifest)
+
+	switch {
+	case bestMatchIsZstd == manifestIsZstd:
+		return false
+	// In any other case, we know the two differ
+	case !didPreferGzip && bestMatchIsZstd:
+		return false // already have an ideal match
+	case !didPreferGzip && manifestIsZstd:
+		return true // a better match
+
+	case didPreferGzip && !bestMatchIsZstd:
+		return false // already have a best match
+	case didPreferGzip && !manifestIsZstd:
+		return true // a better match
+	default:
+		return false
+	}
+}
+
+// instanceIsZstd checks if instance is a zstd instance or not.
+func instanceIsZstd(manifest imgspecv1.Descriptor) bool {
+	if value, ok := manifest.Annotations[OCI1InstanceAnnotationCompressionZSTD]; ok && value == "true" {
+		return true
+	}
+	return false
+}
+
+type instanceCandidate struct {
+	platformIndex    int  // Index of the candidate in platform.WantedPlatforms: lower numbers are preferred; or math.maxInt if the candidate doesn’t have a platform
+	isZstd           bool // or some other expression of the compression, maybe just the annotations hash
+	manifestPosition int  // A zero-based index of the instance in the manifest list
+}
+
+func (ic instanceCandidate) isPreferredOver(other instanceCandidate, preferGzip bool) bool {
+	switch {
+	case ic.platformIndex != other.platformIndex:
+		return ic.platformIndex < other.platformIndex
+	case ic.isZstd != other.isZstd:
+		if !preferGzip {
+			return ic.isZstd
+		} else {
+			return !ic.isZstd
+		}
+		// Or use a compressionScore as suggested elsewhere, for even more regular structure
```

Comment:
Here, as well, either replace the comment with the implementation of that comment, or remove it: it will no be useful to anyone reading the code in the future without referring to this PR.